File: snake/snake_server.py
import sys
from tkinter import *
from random import randrange
import threading
import socket
from snake_common import *
import time

class ClientConnection(threading.Thread):
    def __init__(self, port, action, new_client):
        threading.Thread.__init__(self)
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.sock.bind(("", port))
        self.clients_to_addr = {}
        self.addr_to_clients = {}
        self.min_client_id = 0
        self.action = action
        self.new_client = new_client

    def send(self, message, client):
        self.sock.sendto(message, self.clients_to_addr[client])

    def send_all(self, message):
        for addr in self.addr_to_clients:
            self.sock.sendto(message, addr)

    def run(self):
        while True:
            data, addr = self.sock.recvfrom(4096)
            if addr in self.addr_to_clients:
                self.action(self.addr_to_clients[addr], data)
            else:
                u = self.min_client_id
                self.min_client_id += 1
                self.addr_to_clients[addr] = u
                self.clients_to_addr[u] = addr
                if self.new_client(u):
                    self.action(u, data)
            

class Server():
    def __init__(self):
        self.connection = ClientConnection(SERVER_PORT, self.got_packet, self.new_client)
        self.connection.start()
        self.snakes = {}
        self.directions = {}
        self.snake_colors = {}
        self.apples = []
        self.stopped = False
        self.extend_sizes = {}
        for _ in range(NB_APPLES):
            self.add_apple()
        self.loop()
        
    def set_dir_event(self, d):
        return lambda e: self.set_direction(d)

    def snake_exists(self, p):
        return any(p in snake for snake in self.snakes.values())

    def add_apple(self):
        while True:
            x, y = randrange(WIDTH), randrange(HEIGHT)
            if self.snake_exists((x, y)) or (x, y) in self.apples:
                continue
            self.apples.append((x, y))
            return

    def is_valid(self, p):
        return 0 <= p[0] < WIDTH and 0 <= p[1] < HEIGHT

    def move(self):
        for i in self.snakes:
            snake = self.snakes[i]
            p = snake[-1]
            np = p2add(p, self.directions[i])
            if self.snake_exists(np) or not self.is_valid(np):
                self.stopped = True
                continue
            snake.append(np)
            if np in self.apples:
                self.apples.remove(np)
                self.extend_sizes[i] += 2
                self.add_apple()
            if self.extend_sizes[i] > 0:
                self.extend_sizes[i] -= 1
            else:
                del snake[0]

    def reset(self):
        self.stopped = False
        self.snakes = {}
        self.apples = []
        initial_length = 5
        ahead_length = 3
        w = initial_length + ahead_length
        for cid in self.directions:
            tries = 100
            for _ in range(tries):
                x = randrange(WIDTH - w)
                y = randrange(HEIGHT - w)
                if any(self.snake_exists((x + i, y)) or (x + i, y) in self.apples \
                       for i in range(w)):
                    continue
                self.snakes[cid] = [(x + i, y) for i in range(initial_length)]
                break
            self.directions[cid] = (1, 0)
            self.extend_sizes[cid] = 0
        for _ in range(NB_APPLES):
            self.add_apple()
            

    def loop(self):
        while True:
            if self.stopped:
                time.sleep(2.)
                self.reset()
            self.move()
            self.send_data()
            time.sleep(PERIOD / 1000.)

    def send_data(self):
        for i in self.snakes:
            pck = Packet()
            pck.add_uint8(SET_SNAKE)
            pck.add_uint16(i)
            pck.add_position_list(self.snakes[i])
            self.connection.send_all(pck.data)

        pck = Packet()
        pck.add_uint8(SET_APPLES)
        pck.add_position_list(self.apples)
        self.connection.send_all(pck.data)

    def got_packet(self, client_id, packet):
        pck = Packet(packet)
        command = pck.read_uint8()
        if command == SET_DIRECTION:
            self.directions[client_id] = DIRS[pck.read_uint8()]
        elif command == TOSERVER_INIT:
            rpck = Packet()
            rpck.add_uint8(TOCLIENT_INIT)
            rpck.add_uint16(client_id)
            self.connection.send(rpck.data, client_id)
                
            r, g, b = randrange(1 << 16), randrange(1 << 16), randrange(1 << 16)
            rpck = Packet()
            rpck.add_uint8(SET_SNAKE_COLOR)
            rpck.add_uint16(client_id)
            rpck.add_color((r, g, b))
            self.connection.send_all(rpck.data)

            for i in self.snake_colors:
                rpck = Packet()
                rpck.add_uint8(SET_SNAKE_COLOR)
                rpck.add_uint16(i)
                rpck.add_color(self.snake_colors[i])
                self.connection.send(rpck.data, client_id)

            self.snake_colors[client_id] = (r, g, b)

    def new_client(self, client_id):
        initial_length = 5
        ahead_length = 3
        w = initial_length + ahead_length
        tries = 100
        for _ in range(tries):
            x = randrange(WIDTH - w)
            y = randrange(HEIGHT - w)
            if any(self.snake_exists((x + i, y)) or (x + i, y) in self.apples \
                   for i in range(w)):
                continue
            self.snakes[client_id] = [(x + i, y) for i in range(initial_length)]
            break
        else:
            pck = Packet()
            pck.add_uint8(TOCLIENT_ACCESS_DENIED)
            self.connection.send(pck.data, client_id)
            return False
        self.directions[client_id] = (1, 0)
        self.extend_sizes[client_id] = 0
        print("Hello %d" % client_id)
        return True
        

    def mainloop(self):
        while True:
            pass

    
Server().mainloop()

File: kernel/programs/snake_server/snake_server.cpp
#include "syslib.h"
#include "proglib.h"

u32 rand_state[5];
u32 rand_mods[5] = {131071, 32767, 257, 65535, 2047};
u32 rand_a[5] = {67235, 12345, 195, 20671, 1379};
u32 rand_b[5] = {19237, 3176, 47, 9424, 871};
int rand_next(int n) {
	for (int i = 0; i < 5; i++) {
		rand_state[i] = (rand_state[i] * rand_a[i] + rand_b[i]) % rand_mods[i];
	}
	return rand_state[rand_state[4] % 4] % n;
}
void rand_reseed() {
	void* addr = malloc(1);
	u32 i = (u32)addr;
	free(addr);
	rand_state[0] = i;
	rand_state[1] = i + 17;
	rand_state[2] = i + 42;
	rand_state[3] = i + 51;
	rand_state[4] = i + 23;
	rand_next(1);
	rand_next(1);
	rand_next(1);
}

struct Channel {
	Channel() {}
	Channel(Ipv4Addr addr, uint16_t port) : addr(addr), port(port) {}
	Channel(const UdpSysRead& sysread) :
		addr(sysread.fromAddr), port(sysread.fromPort) {}

	bool operator == (const Channel& oth) const {
		return addr == oth.addr && port == oth.port;
	}

	Ipv4Addr addr;
	uint16_t port;
};
struct ChannelHasher {
	uint32_t operator()(const Channel& ch) const {
		return ch.addr ^ (((uint32_t)ch.port) << 16);
	}
};

typedef HashTable<Channel, int, ChannelHasher> ClientTable;

struct pos {
	int x;
	int y;
	inline pos() {};
	inline pos(int x_, int y_) : x(x_), y(y_) {};
	inline pos operator + (const pos &other) const {
		return pos(x + other.x, y + other.y);
	};
	inline pos operator - (const pos &other) const {
		return pos(x - other.x, y - other.y);
	};
	inline bool operator == (const pos &other) const {
		return (x == other.x) && (y == other.y);
	};
};

struct color {
	uint16_t r;
	uint16_t g;
	uint16_t b;
	color(uint16_t r, uint16_t g, uint16_t b) : r(r), g(g), b(b) {};
	color() {};
};

struct SnakeInfo {
	ExpArray<pos> snake;
	pos direction;
	color snake_color;
	int extend_size;
	SnakeInfo() {
		snake.init();
	}
};

class Packet {
public:
	char *buffer;
	int len;
	int start_pos;
	Packet(char *buffer, int len) :
		buffer(buffer), len(len), start_pos(0) {};
	uint8_t read_uint8() {
		return (uint8_t)buffer[start_pos++];
	};
	void add_uint8(uint8_t v) {
		buffer[len++] = v;
	};
	void add_uint16(uint16_t v) {
		buffer[len++] = (uint8_t)(v >> 8);
		buffer[len++] = (uint8_t)(v & 0xff);
	};
	void add_color(color c) {
		add_uint16(c.r);
		add_uint16(c.g);
		add_uint16(c.b);
	};
	void add_position(pos p) {
		add_uint16(p.x);
		add_uint16(p.y);
	};
	void add_position_list(ExpArray<pos> positions) {
		add_uint16(positions.size());
		for (uint16_t i = 0; i < positions.size(); i++) {
			add_position(positions[i]);
		}
	};
};

enum commands {
	TOSERVER_INIT = 0,
	TOCLIENT_INIT = 1,
	SET_SNAKE = 2,
	SET_APPLES = 3,
	SET_DIRECTION = 4,
	SET_SNAKE_COLOR = 5,
	TOCLIENT_ACCESS_DENIED = 6
};

class SnakeServer {
public:
	const uint16_t LISTEN_PORT = 31412;
	const int16_t WIDTH = 30;
	const int16_t HEIGHT = 30;
	pos DIRS[4];
	const uint16_t NB_APPLES = 3;
	char* send_buffer;
	bool stopped;

	ClientTable clientTable;
	ExpArray<Channel> channelTable;
	int udpSock;

	ExpArray<SnakeInfo> snakes;
	ExpArray<pos> apples;

	SnakeServer() {
		udpSock = udp_bind(LISTEN_PORT);
		if(udpSock < 0)
			exit(3);
		DIRS[0] = pos(0, 1);
		DIRS[1] = pos(1, 0);
		DIRS[2] = pos(-1, 0);
		DIRS[3] = pos(0, -1);
		send_buffer = (char*)malloc(65536);
		stopped = false;
	};

	bool snake_exists(pos p) {
		for (uint16_t i = 0; i < snakes.size(); i++) {
			for (uint16_t j = 0; j < snakes[i].snake.size(); j++) {
				if (snakes[i].snake[j] == p) {
					return true;
				}
			}
		}
		return false;
	};

	bool apple_exists(pos p) {
		for (uint16_t i = 0; i < apples.size(); i++) {
			if (apples[i] == p) {
				return true;
			}
		}
		return false;
	};

	bool is_valid(pos p) {
		return 0 <= p.x && p.x < WIDTH && 0 <= p.y && p.y < HEIGHT;
	};

	void run() {
		while (true) {
			process_incoming();
			move();
			send_data();
			sleep(100 * 1000);
		}
	};

	void move() {
		for (uint16_t i = 0; i < snakes.size(); i++) {
			ExpArray<pos> s = snakes[i].snake;
			pos p = s[s.size() - 1];
			pos np = p + snakes[i].direction;
			if (snake_exists(np) || !is_valid(np)) {
				stopped = true;
				return;
			}
			s.push_back(np);
			if (apple_exists(np)) {
				delete_apple(np);
				snakes[i].extend_size += 2;
				add_apple();
			}
			if (snakes[i].extend_size > 0) {
				snakes[i].extend_size--;
			} else {
				for (uint16_t j = 0; j < s.size() - 1; j++) {
					s[j] = s[j + 1];
				}
				s.pop_back();
			}
		}
	};

	void delete_apple(pos p) {
		for (uint16_t i = 0; i < apples.size(); i++) {
			if (apples[i] == p) {
				apples[i] = apples[apples.size() - 1];
				apples.pop_back();
				return;
			}
		}
	};

	void add_apple() {
		while (true) {
			uint16_t x = rand_next(WIDTH);
			uint16_t y = rand_next(HEIGHT);
			if (snake_exists(pos(x, y)) || apple_exists(pos(x, y))) {
				continue;
			}
			apples.push_back(pos(x, y));
			return;
		}
	};

	void process_packet(int clientId, char* buffer, int len) {
		Packet pck(buffer, len);
		commands cm = (commands)pck.read_uint8();
		if (cm == SET_DIRECTION) {
			snakes[clientId].direction = DIRS[pck.read_uint8()];
		} else if (cm == TOSERVER_INIT) {
			{
				Packet rpck(send_buffer, 0);
				rpck.add_uint8(TOCLIENT_INIT);
				rpck.add_uint16(clientId);
				sendClient(clientId, rpck.buffer, rpck.len);
			}

			uint16_t r = rand_next(1 << 16);
			uint16_t g = rand_next(1 << 16);
			uint16_t b = rand_next(1 << 16);
			snakes[clientId].snake_color = color(r, g, b);

			{
				Packet rpck(send_buffer, 0);
				rpck.add_uint8(SET_SNAKE_COLOR);
				rpck.add_uint16(clientId);
				rpck.add_color(color(r, g, b));
				sendAll(rpck.buffer, rpck.len);
			}
			
			for (uint16_t i = 0; i < snakes.size(); i++) {
				if (i == clientId) continue;
				Packet rpck(send_buffer, 0);
				rpck.add_uint8(SET_SNAKE_COLOR);
				rpck.add_uint16(i);
				rpck.add_color(snakes[i].snake_color);
				sendClient(clientId, rpck.buffer, rpck.len);
			}
		}
	};

	void new_client(int clientId) {
		const int initial_length = 5;
		const int ahead_length = 3;
		const int w = initial_length + ahead_length;
		const int tries = 100;
		for (int _ = 0; _ < tries; _++) {
			uint16_t x = rand_next(WIDTH - w);
			uint16_t y = rand_next(HEIGHT);
			bool ok = true;
			for (int i = 0; i < w; i++) {
				if (snake_exists(pos(x + i, y)) || apple_exists(pos(x + i, y))) {
					ok = false;
					break;
				}
			}
			if (ok) {
				SnakeInfo sn;
				for (int i = 0; i < w; i++) {
					sn.snake.push_back(pos(x + i, y));
				}
				sn.direction = pos(1, 0);
				sn.extend_size = 0;
				snakes.push_back(sn);
				printf("Hello %d\n", clientId);
				return;
			}
		}
		Packet rpck(send_buffer, 0);
		rpck.add_uint8(TOCLIENT_ACCESS_DENIED);
		sendClient(clientId, rpck.buffer, rpck.len);
	};

	void send_data() {
		for (uint16_t i = 0; i < snakes.size(); i++) {
			Packet pck(send_buffer, 0);
			pck.add_uint8(SET_SNAKE);
			pck.add_position_list(snakes[i].snake);
			sendAll(pck.buffer, pck.len);
		}

		Packet pck(send_buffer, 0);
		pck.add_uint8(SET_APPLES);
		pck.add_position_list(apples);
		sendAll(pck.buffer, pck.len);
	};

	void process_incoming() {
		const unsigned BUFFER_SIZE = 2048;
		while(true) {
			UdpSysRead readData;
			char buff[BUFFER_SIZE];
			int len = udp_read(udpSock, (void*)buff, BUFFER_SIZE, &readData);
			if (len == 0) {
				break;
			}

			ClientTable::Iterator itCli =
				clientTable.findIter(Channel(readData));
			
			if(itCli != clientTable.end()) { // Known client
				int clientId = *itCli;
				process_packet(clientId, buff, len);
			} else { // Unknown client
				int clientId = channelTable.size();
				Channel chan(readData);
				clientTable.insert(chan, clientId);
				channelTable.push_back(chan);
				new_client(clientId);
				process_packet(clientId, buff, len);
			}
		}
	};

	void sendClient(int clientId, char* buffer, int size) {
		UdpSysData destDat(channelTable[clientId].addr, LISTEN_PORT,
			channelTable[clientId].port, buffer, size);
		udp_write(&destDat);
	};

	void sendAll(char* buffer, int size) {
		for (uint16_t i = 0; i < channelTable.size(); i++) {
			sendClient(i, buffer, size);
		}
	};
};

int main(int, char**) {
	rand_reseed();
	SnakeServer server;
	server.run();
	return 0;
}
